Download refactor blog post (#3459)
nf-core/website: sites/main-site/src/content/blog/2025/refurbushing-the-pipeline-download.mdx

---
title: "Why are my pipeline download tests failing?"
subtitle: "Refurbishing the pipeline downloads command"
headerImage: "https://images.unsplash.com/photo-1636819488537-a9b1ffb315ce"
headerImageAlt: "3D rendering of a cute little space rocket painted red and white with a fluffy smoke cloud coming out of it."

> pubDate: 2025-08-15T12:00:00+01:00

Reviewer: ```suggestion
pubDate: 2025-08-28T12:00:00+01:00
```

authors:
    - "ErikDanielsson"
    - "mirpedrol"
    - "MatthiasZepper"
label:
    - "tools"
maxHeadingDepth: 3
---

import { Image } from "astro:assets";

import old_nf_version from "@assets/images/blog/pipeline-download-refactor/nf-old-demo.gif";
import docker_download_in_action from "@assets/images/blog/pipeline-download-refactor/bamtofastq-in-action.gif";

The `nf-core pipelines download` command, used when you want to run an `nf-core` pipeline in an offline compute environment, has recently undergone a [substantial refactor](https://github.com/nf-core/tools/pull/3634) that has recently been merged into the development version of `nf-core/tools`.
The following blog post outlines the updates to the command, and explains why your pipeline downloads tests might be failing

### Troubles with container detection and how `nextflow inspect` solves them

> The major problem of the downloads command is to find all containers the pipeline depends on;

Reviewer 2: ````suggestion
The major problem of the downloads command is to find all containers a pipeline depends on;
````

it must do this to be able to bundle the software as a standalone package that can be transferred to the offline machine.
Nextflow allows a user to define container both in modules directly or in config files – as long as the container string is resolved at runtime, Nextflow does not care where it came from.
While this dynamicism makes it possible for Nextflow users to write flexible code, it makes it difficult to determine from the source code which containers the pipeline is using.

> Until now, the download command has solved this problem by building up a code base of complex regex patterns used to search through the pipeline’s files for strings resembling container directives.

Reviewer 2: ````suggestion
Until now, the download command has solved this problem by building up a codebase of complex regex patterns used to search through pipeline files for strings resembling container directives.
````

This workflow checks that the pipeline does not produce any errors when downloaded, to ensure that the pipeline can be used in an offline enviroment.
The workflow is typically only triggered when a PR is made to the pipeline's main branch, i.e. only on release of the pipeline.

The GitHub workflow currently uses the **`dev`** branch of `nf-core/tools`; originally to allow maintainers of `nf-core/tools` to quickly push patches when the regex logic broke.
However, this means that _any_ changes made to the development version of `nf-core/tools` will directly take effect in the GitHub workflow.
Since the refactor of the downloads code requires that the pipeline uses the **25.04** version of Nextflow, pipelines that do not comply with this will fail the download test.

In time, there will be a pipeline template update that will require pipelines to use Nextflow `>=` 25.04, and to thus be compatible with the new command.
The test itself will also be updated to use the `main` branch of `nf-core/tools` to avoid similar issues in the future.

> Until then, you can either voluntarily update the pipelines Nextflow version or ignore the failing test.

Reviewer 2: ```suggestion
Until then, you can either voluntarily update the Nextflow version of your pipelines or ignore the failing test.
```

in addition to Singularity containers which was the only container system previously supported.
The container system to use can be selected via the `--container-system` flag of the command, which now accepts the options `singularity`, `docker` and `none`.
Alternatively, the container system can be selected via the interactive CLI prompts.

The change means that `nf-core` pipelines now can be run on offline HPCs which support only Docker or Podman containers, making it easy to run `nf-core` pipelines in even more compute environments!

:::tip{title="How are Docker images saved?"}

Compared to Singularity containers which are files kept on your file system, Docker images are generally handled by the Docker daemon.
However, via the [`docker image save`](https://docs.docker.com/reference/cli/docker/image/save/) command, Docker allows packaging images as `tar` archives.
The `tar` archives can subsequently be loaded into another Docker daemon with [`docker image load`](https://docs.docker.com/reference/cli/docker/image/load/), or if you are running Podman with [`podman load`](https://docs.podman.io/en/v5.0.2/markdown/podman-load.1.html).

> The `nf-core pipelines downloads` command creates `tar` archives for each Docker image within the pipeline.

Reviewer 2: ````suggestion
The `nf-core pipelines downloads` command creates `tar` archives for each Docker image within the downloaded pipeline.
````

---
title: "Why are my pipeline download tests failing?"
subtitle: "Refurbishing the pipeline downloads command"
headerImage: "https://images.unsplash.com/photo-1636819488537-a9b1ffb315ce"
headerImageAlt: "3D rendering of a cute little space rocket painted red and white with a fluffy smoke cloud coming out of it."
pubDate: 2025-08-15T12:00:00+01:00
authors:
    - "ErikDanielsson"
label:
    - "tools"
maxHeadingDepth: 3
---

import { Image } from "astro:assets";

import old_nf_version from "@assets/images/blog/pipeline-download-refactor/old-nf-version.png";
import rnaseq_in_action from "@assets/images/blog/pipeline-download-refactor/rnaseq-in-action.png";

The `nf-core pipelines download` command, used when you want to run an `nf-core` pipeline in an offline compute environment, has recently undergone a [substantial refactor](https://github.com/nf-core/tools/pull/3634) that has recently been merged into the development version of `nf-core/tools`.
The command has for a long time been reliant on complex regex logic to capture container strings in modules and config files.

> Due to the variability in where a pipeline can define containers, this has been prone to breaking.

Reviewer 3: It would be good to add a paragraph with a bit of history on how the download command used to work befor, and what problems did we find.

Author: Sounds like a good idea!

Reviewer 4: In case you need inspiration...I once made a meme about it.

Good examples of what can go wrong with regexing, [is the quote issue in the Bowtie ](https://github.com/nf-core/tools/issues/2392)module or r[ace conditions because Seqera containers are all called `/data`](https://github.com/nf-core/tools/issues/3285). 

<img width="2760" height="2100" alt="Fixing nf-core download" src="https://github.com/user-attachments/assets/7ddf0e52-6faf-4bba-b36d-ddd147ee9774" />

Author: Thanks! I'll link those PRs too if anyone is interested in learning more. 

Since you have maintained `nf-core download` for such a long time with the regex logic, feel free to update the blog post especially the command history as you see fit!

Not anymore!
As of [Nextflow 25.04](https://www.nextflow.io/docs/latest/migrations/25-04.html) the `nextflow inspect` command has been substantially extended to allow fetching of all pipeline containers by executing a dry-run of the pipeline.
This has now been integrated into the `nf-core/tools` code base which means that the download command can offload the tricky problem of finding pipeline containers to Nextflow, making the command simpler and more reliant at the same time.

> :::warning{title="Why are is the pipeline download tests failing?!"}

Reviewer 3: ```suggestion
:::warning{title="Why are the pipeline download tests failing?!"}
```

Reviewer 3: I think explaining the new command should be explained before this section, and this can be the last one before `further details`

The GitHub workflow currently uses the **`dev`** branch of `nf-core/tools`; originally to allow maintainers of `nf-core/tools` to quickly push patches when the regex logic broke.
However, this means that _any_ changes made to the development version of `nf-core/tools` will directly take effect in the GitHub workflow.
Since the refactor of the downloads code requires that the pipeline uses the **25.04** version of Nextflow, pipelines that do not comply with this will fail the download test.

In time, there will be a pipeline template update that will require pipelines to use Nextflow `>=` 25.04, and to thus be compatible with the new command.
The test itself will also be updated to use the `main` branch of `nf-core/tools` to avoid similar issues in the future.
Until then, you can either voluntarily update the pipelines Nextflow version or ignore the failing test.

:::

### Added support for downloading Docker containers

<Image
    src={rnaseq_in_action}
    class="d-block m-auto"
    width={600}
    density={[1.5, 2]}
    alt="Docker container download for `rnaseq` 3.19.0"
/>

> In addition to the use of `nextflow inspect` for container detection, the download command now also support downloads of Docker container images.

Reviewer 3: ```suggestion
In addition to the use of `nextflow inspect` for container detection, the download command now also supports downloads of Docker container images.
```

> In addition to the use of `nextflow inspect` for container detection, the download command now also support downloads of Docker container images.

Reviewer 3: Add a short sentence saying that the old command downloaded singularity images

The following blog post outlines the updates to the command, and explains why your pipeline downloads tests might be failing

### Troubles with container detection and how `nextflow inspect` solves them

The major problem of the downloads command is to find all containers the pipeline depends on;
it must do this to be able to bundle the software as a standalone package that can be transferred to the offline machine.
Nextflow allows a user to define container both in modules directly or in config files – as long as the container string is resolved at runtime, Nextflow does not care where it came from.
While this dynamicism makes it possible for Nextflow users to write flexible code, it makes it difficult to determine from the source code which containers the pipeline is using.

Until now, the download command has solved this problem by building up a code base of complex regex patterns used to search through the pipeline’s files for strings resembling container directives.
While this strategy has worked for many years in the absence of a better solution, it has been prone to breaking when new edge cases were discovered or when changes were made to the pipeline template.

> However, as of [Nextflow 25.04](https://www.nextflow.io/docs/latest/migrations/25-04.html) the `nextflow inspect` command has been substantially extended to fetch all pipeline containers by executing a dry-run of the pipeline.

Author: Thanks for pointing that out, I must've mixed things up. I'll have a look at the `nextflow inspect` code too, so that I am sure I understand what it does
